### PR TITLE
59 auto linking broken

### PR DESCRIPTION
**Pull request by fxnitsch**
This PR is addressing #59 and
* fixes the linking of dynamically created agents
* adds (better) warning messages if the user provided invalid configurations

Furthermore, it 
* fixes an inconsistency that the `trace_file` is read relatively from the location of the `GeneratorConfig` like all other file paths

Please find a complete configuration for testing purposes attached.
[testing_configuration.zip](https://github.com/user-attachments/files/17024753/testing_scengen_linking_bug.zip)

### CODE REVIEW

**Review by dlr-cjs — CHANGES_REQUESTED** [review]
Please fix issues as discussed

File: scengen/generator.py [hunk 1]
```diff
@@ -83,6 +71,51 @@ def generate_scenario(options: dict) -> None:
...
+        if type_template.get("Contracts"):
+            ids_of_this_agent = [agent["Id"] for agent in scenario["Agents"] if created_agent["this_agent"] in str(agent["Id"])]
+            for this_agent_id in ids_of_this_agent:
+                for external_id in _get_external_ids_from(scenario, created_agent.get("external_ids")):
+                    contracts_to_append = copy.deepcopy(type_template["Contracts"])
+                    _create_contracts(contracts_to_append, this_agent_id, external_id)
+                    scenario["Contracts"].extend(contracts_to_append)
+
+
+def _raise_if_missing_match(create_config: List[Dict]) -> None:
```
[the remark this answers is not in the document]
Comment (fxnitsch) [reply scengen/generator.py]: done

File: scengen/generator.py [hunk 2]
```diff
@@ -55,23 +57,9 @@ def generate_scenario(options: dict) -> None:
...
-                agent_to_append = copy.deepcopy(agent_type_template)
-                agent_to_append["Id"] = _get_agent_id(agent_name, n, n_to_create)
-                scenario["Agents"].append(agent_to_append)
-
-                if contract_type_template:
-                    contract_to_append = copy.deepcopy(contract_type_template)
-                    _replace_ids_in_contracts(contract_to_append, agent_to_append["Id"], external_ids)
-                    scenario["Contracts"].extend(contract_to_append)
+        _raise_if_missing_match(config["create"])
+        add_agents(config, options, scenario)
```
[the remark this answers is not in the document]
Comment (fxnitsch) [reply scengen/generator.py]: done

File: scengen/generator.py [hunk 3]
```diff
@@ -83,6 +71,51 @@ def generate_scenario(options: dict) -> None:
     write_yaml(scenario, options["scenario_path"])
 
 
+def add_agents(config: dict, options: dict, scenario: dict) -> None:
```
[the remark this answers is not in the document]
Comment (fxnitsch) [reply scengen/generator.py]: done

File: scengen/generator.py [hunk 4]
```diff
@@ -83,6 +71,51 @@ def generate_scenario(options: dict) -> None:
...
+    """Adds created agents to `scenario`"""
+    for created_agent in config["create"]:
```
[the remark this answers is not in the document]
Comment (fxnitsch) [reply scengen/generator.py]: done

File: scengen/generator.py [hunk 5]
```diff
@@ -264,28 +297,27 @@ def _get_agent_id(agent_name: str, agent_number: int, n_of_agents_to_create: int
...
-            if isinstance(value, int):
-                replace_map[key] = value
-            elif not value.startswith(REPLACEMENT_IDENTIFIER):
-                replace_map[key] = REPLACEMENT_IDENTIFIER + value
-            else:
-                replace_map[key] = value
-
-    for k, v in replace_map.items():
-        _replace_in_dict(contracts, k, v)
+def _raise_if_no_replacement_identifier_in(contracts: List[Dict]) -> None:
```
[the remark this answers is not in the document]
Comment (fxnitsch) [reply scengen/generator.py]: removed

File: scengen/generator.py [hunk 6]
```diff
@@ -83,6 +71,51 @@ def generate_scenario(options: dict) -> None:
...
+
+        for n in range(n_to_create):
+            agent_to_append = copy.deepcopy(agent_type_template)
+            agent_to_append["Id"] = _get_agent_id(agent_name, n, n_to_create)
+            scenario["Agents"].append(agent_to_append)
+
+
+def add_contracts(config: dict, options: dict, scenario: dict) -> None:
+    """Adds contracts to `scenario`. Note: Ensure that all dynamic agents are already added to the scenario"""
+    for created_agent in config["create"]:
```
[the remark this answers is not in the document]
Comment (fxnitsch) [reply scengen/generator.py]: done

**Comment by fxnitsch** [comment]
@dlr-cjs: all issues resolved. please also find an updated testing configuration in the PR header description.

**Comment by fxnitsch** [comment]
@dlr-cjs: changed `external_ids` to dict and allow list of str/int values, improved pattern matching to be more robust

please review

**Review by dlr-cjs — APPROVED** [review]
Works nicely. generator.py is too large now - I recommend creating a subpackage from it - in a **separate** issue.

**Pull request merged**